### DISCUSSION TITLE
vtsls code action problem

### DISCUSSION THREAD

**Discussion by ffxixslh** [LazyVim/LazyVim#4463] [Q&A] [answered]
Recently, I used LazyVim's default `vtsls` instead of `typescript-tools`, everything seen to be fine.

But once I want to extract selected code with code action like "extract into function/constant",  it should show the rename window first then apply my changes. But it didn't.

It was just jump over the show window part, then show a warning message that said "Language server `vtsls` does not support command `editor.action.rename`. This command may require a client extension.", then leave `newLocal` there. 

How I reproduce the problem:
1. select code
![s1](https://github.com/user-attachments/assets/8625da3e-a4aa-4e69-8ed5-b15a82fc761e)
2. extract into constant
![s2](https://github.com/user-attachments/assets/dba608d7-4c75-421c-b822-29301fb8e6db)
3. show warning message
![s3](https://github.com/user-attachments/assets/d4a066df-0286-4218-802e-c78459945344)

[This](https://github.com/ffxixslh/nvim-config/tree/lazy) is my LazyVim configurations.

Is there any configuration I didn't set or it's a bug?  

And what should I do if I want to use `typescript-tools` instead of `vtsls`?

Highly appreciate if anyone help me solve the problem! Salute！

**Comment by dpetka2001** [comment, marked as answer]
Install [yioneko/nvim-vtsls](https://github.com/yioneko/nvim-vtsls), which seems to be a wrapper around `vtsls` and provides some helper utilities that should be implemented on the client side. It seems to also implement the automatic trigger of rename for inline refactoring. Haven't used it myself, since I don't do web stuff, so you should test it out to see if there are any conflicts.

Otherwise, just disable the Typescript Extra and install `typescript-tools` yourself.

**Reply by ffxixslh** [reply]
Ok, I will try it then. Thank you for your help!

**Reply by agriffis** [reply]
@dpetka2001 Should this become the default for the TypeScript Extra?

**Reply by dpetka2001** [reply]
@agriffis I'm just a simple user like you, who tries to help other people and contribute to the best of my ability and knowledge. I don't do any web dev, so I really am not able to judge if this plugin is needed or not (I actually started learning programming in Lua and Python at the start of this year, so I'm very new to the field of programming).

Someone with knowledge in web dev would qualify better to judge how much improvement this plugin will provide from a user's perspective experience. I just googled around to see if there was a workaround for OP's problem.

If you'd like make a PR and the maintainer will decide.
